highfidelity/hifi
# Avatar Exporter v0.1 - add bone rules and versioning

https://highfidelity.manuscript.com/f/cases/20596
https://highfidelity.manuscript.com/f/cases/20485
https://highfidelity.manuscript.com/f/cases/20612
https://highfidelity.manuscript.com/f/cases/20449
https://highfidelity.manuscript.com/f/cases/20545

Unity Package:
https://github.com/highfidelity/hifi/raw/a9895a90e59c8678a13cdc70be31077bcfe9578b/tools/unity-avatar-exporter/avatarExporter.unitypackage

The current bone rules are:
RecommendedUnityVersion = Warning
SingleRoot = Warning
NoDuplicateMapping = Warning
NoAsymmetricalMapping = Warning
HipsMapped = Error
SpineMapped = Error
SpineDescendantOfHips = Warning
ChestMapped = Error
ChestDescendantOfSpine = Warning
NeckMapped = Warning
HeadMapped = Error
EyesMapped = Warning
HipsNotOnGround = Warning
HipsSpineChestNotCoincident = Warning
TotalBoneCountUnderLimit = Warning

Pre-merge test plan:
- Download the following files:
https://github.com/highfidelity/hifi/raw/a9895a90e59c8678a13cdc70be31077bcfe9578b/tools/unity-avatar-exporter/avatarExporter.unitypackage
https://hifi-content.s3.amazonaws.com/davidback/avatars/fbx/KoolaidMan_Rigged.fbx
https://hifi-content.s3.amazonaws.com/davidback/avatars/fbx/Mr%20Krabs.FBX
https://hifi-content.s3.amazonaws.com/davidback/avatars/fbx/Retsuko_from_blend.fbx
https://hifi-content.s3.amazonaws.com/davidback/avatars/fbx/Tormenta%20White.fbx
https://hifi-content.s3.amazonaws.com/davidback/avatars/fbx/Sara_V1B_3_text.fbx
https://hifi-content.s3.amazonaws.com/davidback/avatars/fbx/03%20-%20robot%20arm%201.fbx
https://hifi-content.s3.amazonaws.com/davidback/avatars/fbx/william.FBX
https://hifi-content.s3.amazonaws.com/davidback/avatars/fbx/Karen_Suit.fbx
https://hifi-content.s3.amazonaws.com/davidback/avatars/fbx/robot_arm_coincident_all.fbx
https://hifi-content.s3.amazonaws.com/davidback/avatars/fbx/skully.fbx
https://hifi-content.s3.amazonaws.com/davidback/avatars/fbx/Chip%20Skylark.fbx
https://hifi-content.s3.amazonaws.com/davidback/avatars/fbx/Mixamo-Alexis_HiRes.fbx
- Ensure you have Unity version 2018.2.12f1 installed, as well as an additional Unity version outside of the 2017.4.17f1 - 2018.2.12f1 version window (latest Unity version will work)
- Open the Unity Version that is not 2018.2.12f1 and open a New Unity Project
- Open the above avatarExporter.unitypackage and import it into your Unity project
- Either select Assets -> Import New Asset and choose any fbx avatar you saved above, or drag the fbx file into the Assets window to import it
- Selec the avatar and, using the Animation Type dropdown in the Rig section of the Inspector on the right, set avatar's Animation Type to Humanoid and hit Apply
- With the avatar still selected, verify when selecting High Fidelity menu > Export New Avatar that a warning is displayed in the Warnings section of the popup window saying you are not using the recommended Unity version
- Close the additional Unity version, open Unity version 2018.2.12f1, and open a new Unity project
- Open the above avatarExporter.unitypackage and import it into your Unity project
- Verify a High Fidelity menu appears as well as a README file in Assets window
- Verify the version 0.1 shows at the top of the README file as well as in the About dialog when choosing High Fidelity menu > About
- Either select Assets -> Import New Asset and choose the fbx avatars you saved above, or drag the fbx files into the Assets window to import it
- Select each avatar and, using the Animation Type dropdown in the Rig section of the Inspector on the right, set each avatar's Animation Type to Humanoid and hit Apply
- Select each avatar, and individually attempt to do a High Fidelity menu > Export New Avatar
- For each avatar, these are the expected results:

KoolaidMan_Rigged - error dialog / cannot export
Errors: No Head bone mapped
Warnings: No Neck bone mapped, no LeftEye/RightEye bone mapped

Mr Krabs - error dialog / cannot export
Errors: No Head bone mapped, no Spine bone mapped, no Chest bone mapped, no Head bone mapped
Warnings: No Neck bone mapped, no LeftEye/RightEye bone mapped

Retsuko - error dialog / cannot export
Errors: No Head bone mapped
Warnings: No Neck bone mapped, no LeftEye/RightEye bone mapped

Tormenta White - Export New Avatar window
Warnings: More than one bone at the top level, no Neck bone mapped, bone mapped to Hips should not be on ground

Sara_V1B_3_text - Export New Avatar window
Warnings: Bone mapped to Hips should not be on ground

03 - robot arm 1 - Export New Avatar window
Warnings: LeftEye bone is mapped to more than one bone

William - Export New Avatar window
Warnings: More than one bone at the top level, nose_L bone is mapped to more than one bone

Karen_Suit - Export New Avatar window
Warnings: Number of bones mapped to left hand doesn't match number mapped to right hand

robot_arm_coincident_all - error dialog / cannot export
Errors: No Chest bone mapped
Warnings: No LeftEye bone mapped

Skully - error dialog / cannot export
Errors: No Chest bone mapped and gives suggestion to map Chest to Chest
Warnings: No LeftEye/RightEye bone mapped

Chip Skylark - error dialog / cannot export
Errors: No Chest bone mapped
Warnings: No Neck bone mapped, no LeftEye bone mapped

Mixamo-Alexis_HiRes - Export New Avatar window
Warnings: None

- Whenever you have the Export New Avatar window opened, verify if clicking Export with an empty Project Name, if typing a Project Name as a directory that already exists at the target location, or if typing an invalid Project Location that a red text error is displayed under Error
- Verify if correcting the errors for either Project Name already existing or Project Location being invalid that the error text changes back to None
- For each of the avatars robot_arm_coincident_all and Chip Skylark, select them and open the Avatar Configuration by choosing Configure in the Rig panel on the right
- Drag the Spine bone from the avatar hierarchy on the left side to the Spine bone mapping on the right side, replacing the existing mapping
- Drag the Chest bone from the avatar hierarchy on the left side to the Chest bone mapping on the right side
- Select Apply then Done
- Verify if attempting to export the avatars again that you get the following results:

robot_arm_coincident_all - Export New Avatar window
Warnings: No LeftEye bone mapped, bones mapped to Hips/Spine/Chest should not be coincidental

Chip Skylark - Export New Avatar window
Warnings: No Neck bone mapped, no LeftEye/RightEye bone mapped

- Continue to export Chip Skylark and Mixamo-Alexis_HiRes, and verify that an Avatar successfully exported dialog appears with the same warnings from above
- For Chip Skylark, make a duplicate of the avatar.fst in the resulting folder
- Open the original avatar.fst in a text editor and delete all of its content
- In Unity with Chip Skylark still selected, choose High Fidelity menu > Update Existing Avatar, and re-export onto the avatar.fst you just edited, and verify that an Avatar successfully exported dialog appears with the same warnings from above
- Upload the resulting project folders to S3, or ask me to upload for you
- Test the resulting fsts (both avatar.fst files for Chip Skylark, an the avatar.fst for Mixamo-Alexis_HiRes) in Interface (any recent master build will suffice)
- Verify the avatars appear correctly in world (besides looking grey since we are not dealing with textures) and animated correctly with movement, handshakes, and emotes
